XenServerHealthCheck/Registry.cs

```diff
@@ -0,0 +1,121 @@
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR 
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using Microsoft.Win32;
+
+namespace XenServerHealthCheck
+{
+    public class Registry
+    {
+        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
+
+        /// <summary>
+        /// Reads a key from XENCENTER_LOCAL_KEYS\k.
+        /// </summary>
+        private static string ReadKey(string k, string def)
+        {
+            try
+            {
+                RegistryKey masterKey = Microsoft.Win32.Registry.LocalMachine.OpenSubKey(HEALTH_CHECK_LOCAL_KEYS);
+                if (masterKey == null)
+                    return def;
+
+                try
+                {
+                    var v = masterKey.GetValue(k);
+                    return (v != null) ? v.ToString() : def;
+                }
+                finally
+                {
+                    masterKey.Close();
+                }
+            }
+            catch (Exception e)
+            {
+                log.DebugFormat(@"Failed to read {0}\{1} from registry; assuming NULL.", HEALTH_CHECK_LOCAL_KEYS, k);
+                log.Debug(e, e);
+                return def;
+            }
+        }
+
+        /// <summary>
+        /// Read a true/false key from XENCENTER_LOCAL_KEYS\k.
+        /// </summary>
+        /// <returns>False if the value is "false" (case insensitive), true if the key is present but not
+        /// "false", def otherwise.</returns>
+        private static Int32 ReadLong(string k, Int32 def)
+        {
+            try
+            {
+                RegistryKey masterKey = Microsoft.Win32.Registry.LocalMachine.OpenSubKey(HEALTH_CHECK_LOCAL_KEYS);
+                if (masterKey == null)
+                    return def;
+
+                try
+                {
+                    object v = masterKey.GetValue(k);
+                    if (v == null)
+                        return def;
+                    else
+                        return Convert.ToInt32(v);
+                }
+                finally
+                {
+                    masterKey.Close();
+                }
+            }
+            catch (Exception e)
+            {
+                log.DebugFormat(@"Failed to read {0}\{1} from registry; assuming {1} is {2}.", HEALTH_CHECK_LOCAL_KEYS, k, def);
+                log.Debug(e, e);
+                return def;
+            }
+        }
+
+        public static string HealthCheckUploadDomainName
+        {
+            get { return ReadKey(HEALTH_CHECK_UPLOAD_DOMAIN_NAME, "https://rttf-staging.citrix.com/feeds/api/"); }

```

Comment:
This could invoke ReadKey and then decide what to do (take result or use a default)